### PR TITLE
OEL-300: Middleware POC.

### PR DESCRIPTION
**Pull request by drishu**

### CODE REVIEW

File: tests/src/Traits/ClientTestTrait.php [hunk 1]
```diff
@@ -6,28 +6,36 @@
...
-        array $responseQueue = [],
-        ?callable $onRequest = null
+        string $responseBody = null
     ): ClientInterface {
-        $mock = new MockHandler($responseQueue, null, null, $onRequest);
-        $handlerStack = HandlerStack::create($mock);
+        $handlerStack = HandlerStack::create();
+        $historyMiddleware = new HistoryMiddleware();
+        $handlerStack->push($historyMiddleware(), 'test.http_client.middleware');
+        $this->historyMiddleware = $historyMiddleware;
```
Comment (drishu) [inline tests/src/Traits/ClientTestTrait.php:34]: this bit here I'm not a fan of but its just a POC

**Review by drishu — COMMENTED** [review]
The branch is red of course, but if you check token and text ingestion are green.
The json changes aren't really needed for the middleware but I've added them in to see how that plays out.

File: tests/src/Api/TextIngestionApiTest.php [hunk 2]
```diff
@@ -45,28 +45,20 @@ public function testTextIngestion(array $clientConfig, array $responses, $expect
...
-            $this->inspectPart($parts[1], 'application/json', 'aclUsers', 20, '["user001","user02"]');
-            $this->inspectPart($parts[2], 'application/json', 'aclNolUsers', 20, '["user003","user04"]');
-            $this->inspectPart($parts[3], 'application/json', 'aclGroups', 23, '["group001","group002"]');
-            $this->inspectPart($parts[4], 'application/json', 'aclNoGroups', 23, '["group003","group004"]');
-            $this->inspectPart($parts[5], 'text/plain', 'text', 43, 'The quick brown fox jumps over the lazy dog');
-        }
+        $request = $historyMiddleware->getHistoryContainer()[0]['request'];
+        $this->inspectTokenRequest($request);
+        $request = $historyMiddleware->getHistoryContainer()[1]['request'];
+        $this->assertEquals('http://web:8080/tests/fixtures/json/ingest-text.json?apiKey=bananas&database=cucumbers&uri=http%3A%2F%2Fexample.com&language=%5B%22en%22%2C%22ro%22%5D&reference=unique-my-ID', $request->getUri());
```
Comment (claudiu-cristea) [inline tests/src/Api/TextIngestionApiTest.php]: We cannot make this assumption. Not everyone is using Docker. I'm not using it, for instance, and I don't have any plan to use it until they'll not fix the macOS performance issue. We need to use a mock as in the previous code. Does the history work with MockHandler? We need to mock the responses and not real HTTP requests.

Comment (drishu) [reply tests/src/Api/TextIngestionApiTest.php]: Indeed as per official docs https://docs.guzzlephp.org/en/stable/testing.html the history middleware and the mock handler should be used together.

File: tests/src/Traits/ClientTestTrait.php [hunk 3]
```diff
@@ -11,23 +11,22 @@
...
     protected function getTestingClient(
         array $configuration = [],
-        array $responseQueue = [],
-        ?callable $onRequest = null
+        HistoryMiddleware $historyMiddleware = null
     ): ClientInterface {
-        $mock = new MockHandler($responseQueue, null, null, $onRequest);
-        $handlerStack = HandlerStack::create($mock);
+        $handlerStack = HandlerStack::create();
+        $handlerStack->push($historyMiddleware(), 'test.http_client.middleware');
```
Comment (brummbar) [inline tests/src/Traits/ClientTestTrait.php]: Just instantiate the middleware here with 
```
Middleware::history($this->clientHistory);
```
Add a property `$clientHistory = []` in the ClientTestTrait.

File: tests/src/HistoryMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+use GuzzleHttp\Middleware;
+
+/**
+ * An HTTP client middleware that wraps the Guzzle history middleware.
+ *
+ * Note: this service has state! It should be only used for test purposes.
+ * It will keep track of the history of requests and responses across all the
+ * HTTP client instances.
+ */
+class HistoryMiddleware
```
Comment (brummbar) [inline tests/src/HistoryMiddleware.php]: This was needed in Drupal, to register as a middleware in the Drupal client stack. But here there's no need for this at all. Also, it's clearly not a service.

File: tests/src/Traits/ClientTestTrait.php (hunk 3 above)
Comment (brummbar) [reply tests/src/Traits/ClientTestTrait.php]: Also drop the second parameter, no need.

File: tests/src/Api/TextIngestionApiTest.php [hunk 5]
```diff
@@ -45,28 +45,20 @@ public function testTextIngestion(array $clientConfig, array $responses, $expect
...
-            $parts = $this->getMultiParts($request);
-            $this->assertCount(6, $parts);
-            $this->inspectPart($parts[0], 'application/json', 'metadata', 55, '{"field1":["value1","value2"],"field2":["value3",2345]}');
-            $this->inspectPart($parts[1], 'application/json', 'aclUsers', 20, '["user001","user02"]');
-            $this->inspectPart($parts[2], 'application/json', 'aclNolUsers', 20, '["user003","user04"]');
-            $this->inspectPart($parts[3], 'application/json', 'aclGroups', 23, '["group001","group002"]');
-            $this->inspectPart($parts[4], 'application/json', 'aclNoGroups', 23, '["group003","group004"]');
-            $this->inspectPart($parts[5], 'text/plain', 'text', 43, 'The quick brown fox jumps over the lazy dog');
-        }
+        $request = $historyMiddleware->getHistoryContainer()[0]['request'];
```
Comment (brummbar) [inline tests/src/Api/TextIngestionApiTest.php]: Always test also the request count, to make sure only the needed ones are performed.

**Review by brummbar — CHANGES_REQUESTED** [review]

File: tests/src/Traits/ClientTestTrait.php [hunk 6]
```diff
@@ -6,28 +6,31 @@
...
      * @return ClientInterface
      */
-    protected function getTestingClient(
-        array $configuration = [],
-        array $responseQueue = [],
-        ?callable $onRequest = null
-    ): ClientInterface {
-        $mock = new MockHandler($responseQueue, null, null, $onRequest);
-        $handlerStack = HandlerStack::create($mock);
+    protected function getTestingClient(array $configuration = [], $responseQueue = []): ClientInterface
```
Comment (claudiu-cristea) [inline tests/src/Traits/ClientTestTrait.php]: Missing `$responseQueue` typing.

File: tests/src/Traits/InspectTestRequestTrait.php [hunk 7]
```diff
@@ -60,9 +60,9 @@ protected function getBoundary(RequestInterface $request): string
      */
     protected function getMultiParts(RequestInterface $request, string $boundary)
     {
-        $parts = explode("--{$boundary}", $request->getBody()->getContents());
+        $parts = explode("--{$this->boundary}", $request->getBody()->__toString());
```
Comment (claudiu-cristea) [inline tests/src/Traits/InspectTestRequestTrait.php]: We are back with `$this->boundary`?

**Review by claudiu-cristea — CHANGES_REQUESTED** [review]

File: tests/src/Traits/ClientTestTrait.php (hunk 3 above)
Comment (drishu) [reply tests/src/Traits/ClientTestTrait.php]: done

File: tests/src/HistoryMiddleware.php (hunk 4 above)
Comment (drishu) [reply tests/src/HistoryMiddleware.php]: done

File: tests/src/Api/TextIngestionApiTest.php (hunk 2 above)
Comment (drishu) [reply tests/src/Api/TextIngestionApiTest.php]: fixed

File: tests/src/Api/TextIngestionApiTest.php (hunk 5 above)
Comment (drishu) [reply tests/src/Api/TextIngestionApiTest.php]: done

**Pull request merged**